WordPress/gutenberg
# Reworked `run` Command Parsing

## What?
This reworks the handling of `wp-env run`'s command parsing in order to more consistently handle quotes and shell tokens.

## Why?
In reworking our E2E setup script I wanted to remove the quotes that https://github.com/WordPress/gutenberg/pull/41179 made unnecessary. When I attempted to do so, however, I found that quotes and environment variables exhibited some very strange behavior. On further examination I found that this was because we were calling `docker-compose` with `shell: true` set. This was causing the command to be processed by two shells before being passed to Docker.

## How?
Originally this pull request only removed the `shell: true` option, but when I did that, it broke _everything else_. The problem I found was that we were heavily relying on the second shell to process the input before running `docker-compose`, and without that shell, we needed to handle that ourselves. The solution here was to just pass the arguments verbatim to Docker, and as a bonus, this made `wp-env` entirely consistent with `docker-compose exec`.

The caveat, however, was that quoted string were now being passed as single arguments (since they are!) instead of separate ones. Rather than splitting the string and making an assumption about what the user was trying to do, I opted for having this as a breaking change. Another problem though: we were relying on quoted strings to support passing options that `wp-env` would otherwise have consumed (like `--help`). The solution I took was to add double dash support since that's pretty standard and already supported out of the box by `yargs`.

The end result is a much more explicit `wp-env run` that does not perform any additional processing on your command. It takes the input it is given (after your local shell processes it) and passes it to Docker verbatim. This resolves any oddities around quoting and a bug wherein you can't pass environment variable tokens to Docker since the local shell will _always_ parse them.

## Testing Instructions

1. Try `npx wp-env run cli bash`.
2. Try `npx wp-env run cli php --help`.
3. Try `npx wp-env run cli php -- --help`.
4. Try `npx wp-env run cli "php"`.
5. Try `npx wp-env run cli "php" \"--help\"`. Notice that it gives an error, but also, that it is passing the quotes as you would expect it to.
